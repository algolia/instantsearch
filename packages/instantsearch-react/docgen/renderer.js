import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
