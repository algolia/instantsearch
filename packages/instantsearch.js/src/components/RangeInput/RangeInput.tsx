/** @jsx h */

import { cx } from '@algolia/ui-components-shared';
import { h, Component } from 'preact';

import Template from '../Template/Template';

import type {
  Range,
  RangeBoundaries,
} from '../../connectors/range/connectRange';
import type { ComponentCSSClasses } from '../../types';
import type {
  RangeInputCSSClasses,
  RangeInputTemplates,
} from '../../widgets/range-input/range-input';

export type RangeInputComponentCSSClasses =
  ComponentCSSClasses<RangeInputCSSClasses>;

export type RangeInputComponentTemplates = Required<RangeInputTemplates>;

export type RangeInputProps = {
  min?: number;
  max?: number;
  step: number;
  values: Partial<Range>;
  cssClasses: RangeInputComponentCSSClasses;
  templateProps: {
    templates: RangeInputComponentTemplates;
  };
  refine: (rangeValue: RangeBoundaries) => void;
};

// Strips leading `0` from a positive number value
function stripLeadingZeroFromInput(value: string): string {
  return value.replace(/^(0+)\d/, (part) => Number(part).toString());
}

class RangeInput extends Component<
  RangeInputProps,
  { min?: string; max?: string }
> {
  public state = {
    min: this.props.values.min?.toString(),
    max: this.props.values.max?.toString(),
  };

  public componentWillReceiveProps(nextProps: RangeInputProps) {
    this.setState({
      min: nextProps.values.min?.toString(),
      max: nextProps.values.max?.toString(),
    });
  }

  private onInput = (key: keyof typeof this.state) => (event: Event) => {
    const { value } = event.currentTarget as HTMLInputElement;

    this.setState({
      [key]: value,
    });
  };

  private onSubmit = (event: Event) => {
    event.preventDefault();

    const { min, max } = this.state;
    this.props.refine([
      min ? Number(min) : undefined,
      max ? Number(max) : undefined,
    ]);
  };

  public render() {
    const { min: minValue, max: maxValue } = this.state;
    const { min, max, step, cssClasses, templateProps } = this.props;
    const isDisabled = min && max ? min >= max : false;
    const hasRefinements = Boolean(minValue || maxValue);
    const rootClassNames = cx(
      cssClasses.root,
      !hasRefinements && cssClasses.noRefinement
    );

    return (
      <div className={rootClassNames}>
        <form className={cssClasses.form} onSubmit={this.onSubmit}>
          <label className={cssClasses.label}>
            <input
              className={cx(cssClasses.input, cssClasses.inputMin)}
              type="number"
              min={min}
              max={max}
              step={step}
              value={stripLeadingZeroFromInput(minValue ?? '')}
              onInput={this.onInput('min')}
              placeholder={min?.toString()}
              disabled={isDisabled}
            />
          </label>

          <Template
            {...templateProps}
            templateKey="separatorText"
            rootTagName="span"
            rootProps={{
              className: cssClasses.separator,
            }}
          />

          <label className={cssClasses.label}>
            <input
              className={cx(cssClasses.input, cssClasses.inputMax)}
              type="number"
              min={min}
              max={max}
              step={step}
              value={stripLeadingZeroFromInput(maxValue ?? '')}
              onInput={this.onInput('max')}
              placeholder={max?.toString()}
              disabled={isDisabled}
            />
          </label>

          <Template
            {...templateProps}
            templateKey="submitText"
            rootTagName="button"
            rootProps={{
              type: 'submit',
              className: cssClasses.submit,
              disabled: isDisabled,
            }}
          />
        </form>
      </div>
    );
  }
}

export default RangeInput;
