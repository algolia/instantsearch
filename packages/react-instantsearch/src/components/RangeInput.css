.root {
}

.labelMin {
}

.inputMin {
}

.labelMax {
}

.inputMax {
}

.submit {
}

.separator {
}
