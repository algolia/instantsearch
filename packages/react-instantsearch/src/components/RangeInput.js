import React, {PropTypes, Component} from 'react';
import {isNaN} from 'lodash';
import translatable from '../core/translatable';
import classNames from './classNames.js';

const cx = classNames('RangeInput');

class RangeInput extends Component {
  static propTypes = {
    translate: PropTypes.func.isRequired,
    refine: PropTypes.func.isRequired,
    min: PropTypes.number,
    max: PropTypes.number,
    currentRefinement: PropTypes.shape({
      min: PropTypes.number,
      max: PropTypes.number,
    }),
    canRefine: PropTypes.bool.isRequired,
  };

  static contextTypes = {
    canRefine: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.state = this.props.canRefine ?
      {from: props.currentRefinement.min, to: props.currentRefinement.max} : {from: '', to: ''};
  }

  componentWillMount() {
    if (this.context.canRefine) this.context.canRefine(this.props.canRefine);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.canRefine) {
      this.setState({from: nextProps.currentRefinement.min, to: nextProps.currentRefinement.max});
    }
    if (this.context.canRefine) this.context.canRefine(nextProps.canRefine);
  }

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();
    if (!isNaN(parseFloat(this.state.from, 10)) && !isNaN(parseFloat(this.state.to, 10))) {
      this.props.refine({min: this.state.from, max: this.state.to});
    }
  };

  render() {
    const {translate, canRefine} = this.props;
    return (
      <form {...cx('root', !canRefine && 'noRefinement')} onSubmit={this.onSubmit}>
        <label {...cx('labelMin')}>
          <input {...cx('inputMin')}
                 type="number" value={this.state.from} onChange={e => this.setState({from: e.target.value})}
          />
        </label>
        <span {...cx('separator')}>{translate('separator')}</span>
        <label {...cx('labelMax')}>
          <input {...cx('inputMax')}
                 type="number" value={this.state.to} onChange={e => this.setState({to: e.target.value})}
          />
        </label>
        <button {...cx('submit')} type="submit">{translate('submit')}
        </button>
      </form>
    );
  }
}

export default translatable({
  submit: 'ok',
  separator: 'to',
})(RangeInput);
