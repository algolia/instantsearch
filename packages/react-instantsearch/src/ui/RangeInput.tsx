import React, { useState } from 'react';

import { cx } from './lib/cx';

import type { useRange } from 'react-instantsearch-core';

type RangeRenderState = ReturnType<typeof useRange>;

export type RangeInputProps = Omit<React.ComponentProps<'div'>, 'onSubmit'> &
  Pick<RangeRenderState, 'range' | 'start'> & {
    classNames?: Partial<RangeInputClassNames>;
    disabled: boolean;
    onSubmit: RangeRenderState['refine'];
    step?: number;
    translations: RangeInputTranslations;
  };

export type RangeInputClassNames = {
  /**
   * Class names to apply to the root element
   */
  root: string;
  /**
   * Class names to apply to the root element when there are no refinements possible
   */
  noRefinementRoot: string;
  /**
   * Class names to apply to the form element
   */
  form: string;
  /**
   * Class names to apply to each label element
   */
  label: string;
  /**
   * Class names to apply to each input element
   */
  input: string;
  /**
   * Class names to apply to the minimum input element
   */
  inputMin: string;
  /**
   * Class names to apply to the maximum input element
   */
  inputMax: string;
  /**
   * Class names to apply to the separator element
   */
  separator: string;
  /**
   * Class names to apply to the submit button
   */
  submit: string;
};

export type RangeInputTranslations = {
  /**
   * The label of the separator, between the minimum and maximum inputs
   */
  separatorElementText: string;
  /**
   * The label of the submit button
   */
  submitButtonText: string;
};

// if the default value is undefined, React considers the component uncontrolled initially, which we don't want 0 or NaN as the default value
const unsetNumberInputValue = '';

// Strips leading `0` from a positive number value
function stripLeadingZeroFromInput(value: string): string {
  return value.replace(/^(0+)\d/, (part) => Number(part).toString());
}

export function RangeInput({
  classNames = {},
  range: { min, max },
  start: [minValue, maxValue],
  step = 1,
  disabled,
  onSubmit,
  translations,
  ...props
}: RangeInputProps) {
  const values = {
    min:
      minValue !== -Infinity && minValue !== min
        ? minValue
        : unsetNumberInputValue,
    max:
      maxValue !== Infinity && maxValue !== max
        ? maxValue
        : unsetNumberInputValue,
  };
  const [prevValues, setPrevValues] = useState(values);

  const [{ from, to }, setRange] = useState({
    from: values.min?.toString(),
    to: values.max?.toString(),
  });

  if (values.min !== prevValues.min || values.max !== prevValues.max) {
    setRange({ from: values.min?.toString(), to: values.max?.toString() });
    setPrevValues(values);
  }

  return (
    <div
      {...props}
      className={cx(
        cx('ais-RangeInput', classNames.root),
        disabled &&
          cx('ais-RangeInput--noRefinement', classNames.noRefinementRoot),
        props.className
      )}
    >
      <form
        className={cx('ais-RangeInput-form', classNames.form)}
        onSubmit={(event) => {
          event.preventDefault();
          onSubmit([
            from ? Number(from) : undefined,
            to ? Number(to) : undefined,
          ]);
        }}
      >
        <label className={cx('ais-RangeInput-label', classNames.label)}>
          <input
            className={cx(
              'ais-RangeInput-input',
              classNames.input,
              'ais-RangeInput-input--min',
              classNames.inputMin
            )}
            type="number"
            min={min}
            max={max}
            value={stripLeadingZeroFromInput(from || unsetNumberInputValue)}
            step={step}
            placeholder={min?.toString()}
            disabled={disabled}
            onInput={({ currentTarget }) => {
              const value = currentTarget.value;
              setRange({
                from: value || unsetNumberInputValue,
                to,
              });
            }}
          />
        </label>
        <span className={cx('ais-RangeInput-separator', classNames.separator)}>
          {translations.separatorElementText}
        </span>
        <label className={cx('ais-RangeInput-label', classNames.label)}>
          <input
            className={cx(
              'ais-RangeInput-input',
              classNames.input,
              'ais-RangeInput-input--max',
              classNames.inputMax
            )}
            type="number"
            min={min}
            max={max}
            value={stripLeadingZeroFromInput(to || unsetNumberInputValue)}
            step={step}
            placeholder={max?.toString()}
            disabled={disabled}
            onInput={({ currentTarget }) => {
              const value = currentTarget.value;
              setRange({
                from,
                to: value || unsetNumberInputValue,
              });
            }}
          />
        </label>
        <button
          className={cx('ais-RangeInput-submit', classNames.submit)}
          type="submit"
        >
          {translations.submitButtonText}
        </button>
      </form>
    </div>
  );
}
