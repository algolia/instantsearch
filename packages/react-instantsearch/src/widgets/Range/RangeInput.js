import React, {PropTypes, Component} from 'react';

import themeable from '../../core/themeable';
import {isNaN} from 'lodash';
import translatable from '../../core/translatable';

import theme from './RangeInput.css';

class RangeInput extends Component {
  static propTypes = {
    applyTheme: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    refine: PropTypes.func.isRequired,
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    value: PropTypes.shape({
      min: PropTypes.number,
      max: PropTypes.number,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {from: props.value.min, to: props.value.max};
  }

  componentWillReceiveProps(nextProps) {
    this.setState({from: nextProps.value.min, to: nextProps.value.max});
  }

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();
    if (!isNaN(parseFloat(this.state.from, 10)) && !isNaN(parseFloat(this.state.to, 10))) {
      this.props.refine({min: this.state.from, max: this.state.to});
    }
  };

  render() {
    const {applyTheme, translate} = this.props;
    return (
      <form {...applyTheme('root', 'root')} onSubmit={this.onSubmit}>
        <label {...applyTheme('labelMin', 'labelMin')}>
          <input {...applyTheme('inputMin', 'inputMin')}
                 type="number" value={this.state.from} onChange={e => this.setState({from: e.target.value})}
          />
        </label>
        <span {...applyTheme('separator', 'separator')}>{translate('separator')}</span>
        <label {...applyTheme('labelMax', 'labelMax')}>
          <input {...applyTheme('inputMax', 'inputMax')}
                 type="number" value={this.state.to} onChange={e => this.setState({to: e.target.value})}
          />
        </label>
        <button {...applyTheme('submit', 'submit')} type="submit">{translate('submit')}
        </button>
      </form>
    );
  }
}

export default themeable(theme)(translatable({
  submit: 'go',
  separator: 'to',
})(RangeInput)
);
