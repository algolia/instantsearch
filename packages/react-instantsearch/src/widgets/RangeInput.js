import connectRange from '../connectors/connectRange.js';
import RangeInputComponent from '../components/RangeInput.js';

/**
 * RangeInput is a widget that allows a user to select a numeric range using inputs.
 * @name RangeInput
 * @kind component
 * @category widget
 * @propType {string} attributeName - the name of the attribute in the record
 * @propType {{min: number, max: number}} defaultRefinement - Default state of the widget containing the start and the end of the range.
 * @propType {number} min - Minimum value. When this isn't set, the minimum value will be automatically computed by Algolia using the data in the index.
 * @propType {number} max - Maximum value. When this isn't set, the maximum value will be automatically computed by Algolia using the data in the index.
 * @themeKey ais-RangeInput__root - The root component of the widget
 * @themeKey ais-RangeInput__labelMin - The label for the min input
 * @themeKey ais-RangeInput__inputMin - The min input
 * @themeKey ais-RangeInput__separator - The separator between input
 * @themeKey ais-RangeInput__labelMax - The label for the max input
 * @themeKey ais-RangeInput__inputMax - The max input
 * @themeKey ais-RangeInput__submit - The submit button
 * @translationKey submit - Label value for the submit button
 * @translationKey separator - Label value for the input separator
 * @example
 * import React from 'react';
 *
 * import {RangeInput, InstantSearch} from '../packages/react-instantsearch/dom';
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       className="container-fluid"
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *        <RangeInput attributeName="price"/>
 *     </InstantSearch>
 *   );
 * }
 */
export default connectRange(RangeInputComponent);
