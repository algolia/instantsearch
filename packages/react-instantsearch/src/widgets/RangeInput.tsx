import React from 'react';
import { useRange } from 'react-instantsearch-core';

import { RangeInput as RangeInputUiComponent } from '../ui/RangeInput';

import type { RangeInputProps as RangeInputUiProps } from '../ui/RangeInput';
import type { UseRangeProps } from 'react-instantsearch-core';

type UiProps = Pick<
  RangeInputUiProps,
  'disabled' | 'onSubmit' | 'range' | 'start' | 'step' | 'translations'
>;

export type RangeInputProps = Omit<RangeInputUiProps, keyof UiProps> &
  UseRangeProps & { translations?: Partial<UiProps['translations']> };

export function RangeInput({
  attribute,
  min,
  max,
  precision,
  translations,
  ...props
}: RangeInputProps) {
  const { range, start, canRefine, refine } = useRange(
    { attribute, min, max, precision },
    { $$widgetType: 'ais.rangeInput' }
  );

  const step = 1 / Math.pow(10, precision || 0);

  const uiProps: UiProps = {
    range,
    start,
    step,
    disabled: !canRefine,
    onSubmit: refine,
    translations: {
      separatorElementText: 'to',
      submitButtonText: 'Go',
      ...translations,
    },
  };

  return <RangeInputUiComponent {...props} {...uiProps} />;
}
