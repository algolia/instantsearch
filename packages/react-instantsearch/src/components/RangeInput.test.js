/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

import RangeInput from './RangeInput';

describe('RangeInput', () => {
  it('supports passing max/min values', () => {
    const tree = renderer.create(
      <RangeInput
        createURL={() => '#'}
        refine={() => null}
        min={0}
        max={100}
        currentRefinement={{min: 0, max: 100}}
        canRefine={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies translations', () => {
    const tree = renderer.create(
      <RangeInput
        createURL={() => '#'}
        refine={() => null}
        translations={{
          submit: 'SUBMIT',
          separator: 'SEPARATOR',
        }}
        min={0}
        max={100}
        currentRefinement={{min: 0, max: 100}}
        canRefine={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(
        <RangeInput
          createURL={() => '#'}
          refine={refine}
          min={0}
          max={100}
          currentRefinement={{min: 0, max: 100}}
          canRefine={true}
        />
      );

    const formChildren = wrapper.find('.ais-RangeInput__submit');
    wrapper
        .find('.ais-RangeInput__root')
        .simulate('submit', {target: {formChildren}});

    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toEqual({min: 0, max: 100});

    refine.mockClear();

    wrapper.find('input').first().simulate('change', {target: {value: 89}});
    wrapper.find('input').last().simulate('change', {target: {value: 99}});

    wrapper
        .find('.ais-RangeInput__root')
        .simulate('submit', {target: {formChildren}});

    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toEqual({min: 89, max: 99});

    wrapper.unmount();
  });

  it('do not refine where input value are empty string', () => {
    const refine = jest.fn();
    const wrapper = mount(
        <RangeInput
          createURL={() => '#'}
          refine={refine}
          min={0}
          max={100}
          currentRefinement={{min: 0, max: 100}}
          canRefine={true}
        />
      );

    const formChildren = wrapper.find('.ais-RangeInput__submit');

    wrapper.find('input').first().simulate('change', {target: {value: ''}});
    wrapper.find('input').last().simulate('change', {target: {value: ''}});

    wrapper
        .find('.ais-RangeInput__root')
        .simulate('submit', {target: {formChildren}});

    expect(refine.mock.calls.length).toBe(0);

    wrapper.unmount();
  });

  describe('Panel compatibility', () => {
    it('Should indicate when no more refinement', () => {
      const canRefine = jest.fn();
      const wrapper = mount(
        <RangeInput
          createURL={() => '#'}
          refine={() => {}}
          min={0}
          max={100}
          currentRefinement={{min: 0, max: 100}}
          canRefine={true}
        />,
        {
          context: {canRefine},
          childContextTypes: {canRefine: React.PropTypes.func},
        },
      );

      expect(canRefine.mock.calls.length).toBe(1);
      expect(canRefine.mock.calls[0][0]).toEqual(true);
      expect(wrapper.find('.ais-RangeInput__noRefinement').length).toBe(0);

      wrapper.setProps({canRefine: false});

      expect(canRefine.mock.calls.length).toBe(2);
      expect(canRefine.mock.calls[1][0]).toEqual(false);
      expect(wrapper.find('.ais-RangeInput__noRefinement').length).toBe(1);
    });
  });
});
