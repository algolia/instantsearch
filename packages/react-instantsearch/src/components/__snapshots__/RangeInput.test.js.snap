exports[`RangeInput applies translations 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}>
  <label
    className="ais-RangeInput__labelMin">
    <input
      className="ais-RangeInput__inputMin"
      onChange={[Function]}
      type="number"
      value={0} />
  </label>
  <span
    className="ais-RangeInput__separator">
    SEPARATOR
  </span>
  <label
    className="ais-RangeInput__labelMax">
    <input
      className="ais-RangeInput__inputMax"
      onChange={[Function]}
      type="number"
      value={100} />
  </label>
  <button
    className="ais-RangeInput__submit"
    type="submit">
    SUBMIT
  </button>
</form>
`;

exports[`RangeInput supports passing max/min values 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}>
  <label
    className="ais-RangeInput__labelMin">
    <input
      className="ais-RangeInput__inputMin"
      onChange={[Function]}
      type="number"
      value={0} />
  </label>
  <span
    className="ais-RangeInput__separator">
    to
  </span>
  <label
    className="ais-RangeInput__labelMax">
    <input
      className="ais-RangeInput__inputMax"
      onChange={[Function]}
      type="number"
      value={100} />
  </label>
  <button
    className="ais-RangeInput__submit"
    type="submit">
    ok
  </button>
</form>
`;
